dbms/src/Storages/Transaction/RegionPersister.cpp

```diff
@@ -247,17 +286,36 @@ RegionMap RegionPersister::restore(const TiFlashRaftProxyHelper * proxy_helper,
             page_storage_v2->restore();
             page_storage_v3->restore();
 
-            if (page_storage_v2->getNumberOfPages() == 0)
-            {
+            auto change_to_only_v3 = [&]() {
                 page_storage_v2 = nullptr;
-                run_mode = PageStorageRunMode::ONLY_V3;
                 page_writer = std::make_shared<PageWriter>(run_mode, /*storage_v2_*/ nullptr, page_storage_v3);
                 page_reader = std::make_shared<PageReader>(run_mode, ns_id, /*storage_v2_*/ nullptr, page_storage_v3, global_context.getReadLimiter());
-            }
-            else
+
+                run_mode = PageStorageRunMode::ONLY_V3;
+            };
+
+            if (const auto & kvstore_remain_pages = page_storage_v2->getNumberOfPages(); kvstore_remain_pages != 0)
             {
                 page_writer = std::make_shared<PageWriter>(run_mode, page_storage_v2, page_storage_v3);
                 page_reader = std::make_shared<PageReader>(run_mode, ns_id, page_storage_v2, page_storage_v3, global_context.getReadLimiter());
+
+                forceTransformKVStoreV2toV3();
+                const auto & kvstore_remain_pages_after_transform = page_storage_v2->getNumberOfPages();
+                LOG_FMT_INFO(log, "Current kvstore translate to V3 finished. [ns_id={}] [done={}] [remain_before_translate_pages={}], [remain_after_translate_pages={}]", //
+                             ns_id,
+                             kvstore_remain_pages_after_transform == 0,
+                             kvstore_remain_pages,
+                             kvstore_remain_pages_after_transform);
+
+                if (kvstore_remain_pages_after_transform == 0)
```

Comment:
Under which case will `kvstore_remain_pages_after_transform` not be 0 after `forceTransformKVStoreV2toV3`?

Reply:
A logical error may cause that.

Will change it to `if (kvstore_remain_pages_after_transform != 0) throw`